lucene/core/src/java/org/apache/lucene/util/hnsw/BoundsChecker.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.util.hnsw;
+
+abstract class BoundsChecker {
+
+    float bound;
+
+    /**
+     * Update the bound if sample is better
+     */
+    abstract void update(float sample);
+
+    /**
+     * Return whether the sample exceeds (is worse than) the bound
+     */
+    abstract boolean check(float sample);
+
+    static BoundsChecker create(boolean reversed) {
+        if (reversed) {
+            return new Min();
+        } else {
+            return new Max();
+        }
+    }
+
+    static class Max extends BoundsChecker {
+        Max() {
+            bound = -Float.MAX_VALUE;
```

Comment:
At first I thought we should use `Float.MIN_VALUE` here, but that is not right, since it is the smallest positive float.

But then I worried whether negating `Float.MAX_VALUE` produces a valid float (e.g. you cannot safely negate `Integer.MIN_VALUE`).

Maybe just use `Float.NEGATIVE_INFINITY` here and `Float.POSITIVE_INFINITY` for `Min`?

Reply:
Well, `-Float.MAX_VALUE` is reliable since the sign bit is separate from all the rest of the bits, but `Float.NEGATIVE_INFINITY` is less than that, so if we use it then any finite float will be a valid score, so +1 for that, and for +infinity as well